meshmode/dof_array.py

```diff
@@ -390,26 +391,45 @@ def _flatten_dof_array(ary: Any, strict: bool = True):
     if actx is None:
         raise ValueError("cannot flatten frozen DOFArrays")
 
-    @memoize_in(actx, (_flatten_dof_array, "flatten_prg"))
+    @memoize_in(actx, (_flatten_dof_array, "flatten_grp_ary_prg"))
     def prg():
         return make_loopy_program(
-            "{[iel,idof]: 0<=iel<nelements and 0<=idof<ndofs_per_element}",
-            """result[grp_start + iel*ndofs_per_element + idof] \
-                = grp_ary[iel, idof]""",
-            name="flatten")
-
-    group_sizes = [grp_ary.shape[0] * grp_ary.shape[1] for grp_ary in ary]
-    group_starts = np.cumsum([0] + group_sizes)
-
-    result = actx.empty(group_starts[-1], dtype=ary.entry_dtype)
-
-    for grp_start, grp_ary in zip(group_starts, ary):
-        actx.call_loopy(prg(),
-                grp_ary=grp_ary,
-                result=result,
-                grp_start=grp_start)
+            [
+                "{[iel]: 0 <= iel < nelements}",
+                "{[idof]: 0 <= idof < ndofs_per_element}"
+            ],
+            """
+                result[iel * ndofs_per_element + idof] = grp_ary[iel, idof]
+            """,
+            [
+                lp.GlobalArg("result", None,
+                             shape="nelements * ndofs_per_element"),
+                lp.GlobalArg("grp_ary", None,
+                             shape=("nelements", "ndofs_per_element")),
+                lp.ValueArg("nelements", np.int32),
+                lp.ValueArg("ndofs_per_element", np.int32),
+                "..."
+            ],
+            name="flatten_grp_ary"
+        )
+
+    def _flatten(grp_ary):
+        if grp_ary.flags.c_contiguous:
+            return grp_ary.reshape(-1, order="C")
+        elif grp_ary.flags.f_contiguous:
+            return grp_ary.reshape(-1, order="F")
+        else:
+            # NOTE: array has unsupported strides
+            return actx.call_loopy(
+                prg(),
+                grp_ary=grp_ary
+            )["result"]
 
-    return result
+    if len(ary) == 1:
+        # Special case which avoids a copy
+        return _flatten(ary[0])
```

Comment:
```suggestion
        # can avoid a copy if reshape succeeds
        return _flatten(ary[0])
```
?

Do you think the same special case should be added to `unflatten`? That way it has the potential of doing a zero-copy roundtrip. I don't know of any place that would actually make use of that though.

Reply:
Maybe? I'm a bit hesitant to touch `unflatten` and there are no stateful updates going on there. @inducer?

Reply:
Fair point! I'm fine with leaving it as is.